src/api/json_tokenize.cpp

```diff
@@ -0,0 +1,116 @@
+#include "json_tokenize.hpp"
+#include <algorithm>
+#include <string>
+#include <vector>
+#include <iostream>
+#include <migraphx/errors.hpp>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+siter colon(siter start, siter end)
+{
+    return std::find_if(start, end, [](auto c) { return c == ':'; });
+}
+
+std::pair<siter, siter> key(siter start, siter end)
+{
+    // find key end
+    --end;
+    while(start != end)
+    {
+        if(*end != ' ')
+        {
+            break;
+        }
+        --end;
+    }
+    auto ke = end;
+
+    if(start == end)
+    {
+        if(*ke != '\"')
+        {
+            return {ke, ke};
+        }
+        else
+        {
+            MIGRAPHX_THROW("KEY: single quote cannot be a key!");
+        }
+    }
+
+    // find key start
+    --end;
+    while(start != end)
+    {
+        // match
+        if(*ke == '\"' and *end == '\"')
+        {
+            break;
+        }
+        else if(*ke != '\"' and (std::ispunct(*end) or *end == ' '))
+        {
+            ++end;
+            break;
+        }
+        --end;
+    }
```

Comment:
This loop could be written as this:

```cpp
if(*ke == '\"')
{
    end = std::find_if(std::make_reverse_iterator(end), std::make_reverse_iterator(start), [](char c) {
        return c == '\"';
    }).base();
}
else
{
    end = std::find_if(std::make_reverse_iterator(end), std::make_reverse_iterator(start), [](char c) {
        return std::ispunct(c) or c == ' ';
    }).base();
    end++;
}
```

The `*ke` condition is moved out of the loop, and we use `std::find_if`.